scripts/activation_patching_ioi.py

```diff
@@ -0,0 +1,104 @@
+"""
+
+Use activation patching to study indirect object identification (IOI) on gpt2-small
+
+The task in IOI is to identify the correct indirect object in a sentence like:
+    "After John and Mary went to the store, Mary gave a bottle of milk to"
+
+With activation patching, we can analyze which parts of a model contribute to the task.
+
+Resources:
+Interpretability in the wild paper: https://arxiv.org/pdf/2211.00593.pdf
+Neel Nandas Activation patching with TL Colab:
+    https://colab.research.google.com/github/neelnanda-io/TransformerLens/blob/main/demos/Activation_Patching_in_TL_Demo.ipynb
+
+In this script, we implement activation patching via the patchscope framework.
+
+"""
+
+from obvs.patchscope import SourceContext, TargetContext, Patchscope
+from obvs.vis import create_heatmap
+
+
+# define metric
+def ioi_metric(patched_logits, clean_logits, corrupt_logits, correct_idx, incorrect_idx):
+    """ Metric for checking correctness of indirect object identification
+        the normalized_logit_diff is constructed so that it is 1, if the output logits are the
+        same as in the clean run, and 0 if the output logits are the same as in the corrupted run
+        it increases, if the patched activation contribute in making the output logits
+        more like in the clean run
+    """
+
+    # get the difference in logits of the correct token and the incorrect token at the last
+    # position
+    patched_logit_diff = (patched_logits[-1, correct_idx] - patched_logits[-1, incorrect_idx])
+    clean_logit_diff = (clean_logits[-1, correct_idx] - clean_logits[-1, incorrect_idx])
+    corrupt_logit_diff = (corrupt_logits[-1, correct_idx] - corrupt_logits[-1, incorrect_idx])
+
+    return (patched_logit_diff - corrupt_logit_diff) / (clean_logit_diff - corrupt_logit_diff)
+
+# setup
+# the clean prompt produces our baseline answer, the corrupted prompt will be patched with
+# activations from the clean prompt run later on
+clean_prompt = "After John and Mary went to the store, Mary gave a bottle of milk to"
+corrupted_prompt = "After John and Mary went to the store, John gave a bottle of milk to"
+
+# setup patchscope
+source_context = SourceContext(prompt=clean_prompt, model_name='gpt2')
+target_context = TargetContext(prompt=corrupted_prompt, model_name='gpt2', max_new_tokens=1)
+patchscope = Patchscope(source_context, target_context)
+
+
+# get the indices of the correct and incorrect tokens
+correct_index = patchscope.tokenizer(" John")["input_ids"][0]
+incorrect_index = patchscope.tokenizer(" Mary")["input_ids"][0]
+
+
+# first, do a forward pass on the corrupted prompt to get the corrupted logits
+patchscope.source.prompt = corrupted_prompt
+patchscope.source_forward_pass()
+corrupted_logits = patchscope.source_output.detach()
+
+
+# then, do a forward pass on the clean prompt to get the clean logits
+patchscope.source.prompt = clean_prompt
+patchscope.source_forward_pass()
+clean_logits = patchscope.source_output.detach()
```

Comment:
Do we need to do this hack to get the correct results? Can we use a single `.run()` to get logits from source and target?

Reply:
Just noticed that nnsight supports batching with multiple `invoke` contexts

```
with model.trace() as tracer:

  with tracer.invoke('The Eiffel Tower is in the city of'):

    # Ablate the last MLP for only this batch.
    model.transformer.h[-1].mlp.output[0][:] = 0

    # Get the output for only the intervened on batch.
    token_ids_intervention = model.lm_head.output.argmax(dim=-1).save()

  with tracer.invoke('The Eiffel Tower is in the city of'):

    # Get the output for only the original batch.
    token_ids_original = model.lm_head.output.argmax(dim=-1).save()

print("Original token IDs:", token_ids_original)
print("Intervention token IDs:", token_ids_intervention)

print("Original prediction:", model.tokenizer.decode(token_ids_original[0][-1]))
print("Intervention prediction:", model.tokenizer.decode(token_ids_intervention[0][-1]))
```
https://nnsight.net/notebooks/tutorials/walkthrough/

I think patchscope should also support batching. That way, we can get all the results here with a single forward pass.